### PR TITLE
Review

### PR DESCRIPTION
**Pull request by bl0rch1d**

### CODE REVIEW

File: app/controllers/app_controller.rb [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+
+    @session.save(player: @player, game: @game)
+
+    return redirect_to :game_results if game_finished?
+
+    show_page :game
+  end
+
+  def hint
+    return redirect_to :root unless @session.present?
```
Comment (yaorlov) [inline app/controllers/app_controller.rb]: Let's add a custom rack middleware for checking if the user is authorized and remove these duplications:
`return redirect_to :root unless @session.present?`
The middleware should return 403 status

**Comment by bl0rch1d** [comment]
There is one problem with RSpec. He doesn't see my middleware and tests are falling. On server all works fine.

Please, HALP!

**Comment by yaorlov** [comment]
@bl0rch1d which tests are failing? I see that your tests passed on CI

**Comment by bl0rch1d** [comment]
> @bl0rch1d which tests are failing? I see that your tests passed on CI

I commented them

https://github.com/bl0rch1d/codebreaker_rack/blob/develop/spec/codebreaker_rack/codebreaker_rack_spec.rb#L96

https://github.com/bl0rch1d/codebreaker_rack/blob/develop/spec/codebreaker_rack/codebreaker_rack_spec.rb#L120

File: app/controllers/base_controller.rb [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    PATHS.each_with_index do |path, index|
+      Dir.entries(path).each do |file|
+        views[views.keys[index]][file.split('.').first.intern] = file if file.match?(regex)
+      end
+    end
+
+    views
+  end
+
+  VIEWS ||= find_views
```
Comment (yaorlov) [inline app/controllers/base_controller.rb]: move it to `views` method with memoization: https://www.justinweiss.com/articles/4-simple-memoization-patterns-in-ruby-and-one-gem/

File: app/middleware/auth_redirect_middleware.rb [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+    def initialize(app, status = 302)
+      @app = app
+      @status = status
+    end
+
+    def call(env)
+      @request = Rack::Request.new(env)
+
+      return [@status, { 'Location' => '/' }, ['']] if !authenticated? && auth_location?
```
Comment (yaorlov) [inline app/middleware/auth_redirect_middleware.rb]: replace magic strings with constants: `Router::URLS[:root]`

File: app/controllers/base_controller.rb (hunk 2 above)
Comment (bl0rch1d) [reply app/controllers/base_controller.rb]: image
Is that correct?

File: app/middleware/auth_redirect_middleware.rb [hunk 4]
```diff
@@ -0,0 +1,41 @@
+# frozen_string_literal: true
+
+module Rack
```
Comment (yaorlov) [inline app/middleware/auth_redirect_middleware.rb]: You have invalid naming. According to your directory and file name, it should be `Middleware::AuthRedirectMiddleware`.
Please check zeitwerk gem documentation for more examples.
P.S. custom code like this middleware is usually located in the `lib` folder, not in the `middleware`

Comment (bl0rch1d) [reply app/middleware/auth_redirect_middleware.rb]: ok

File: app/helpers/session_helper.rb [hunk 5]
```diff
@@ -0,0 +1,27 @@
+# frozen_string_literal: true
+
+class SessionHelper
```
Comment (yaorlov) [inline app/helpers/session_helper.rb]: Invalid naming
```
module Helpers
  class SessionHelper
```

File: app/controllers/base_controller.rb [hunk 6]
```diff
@@ -0,0 +1,65 @@
+# frozen_string_literal: true
+
+class BaseController
```
Comment (yaorlov) [inline app/controllers/base_controller.rb]: `module Controllers`

File: app/middleware/auth_redirect_middleware.rb (hunk 4 above)
Comment (bl0rch1d) [reply app/middleware/auth_redirect_middleware.rb]: My middleware must be located in root of lib/ directory?

Comment (yaorlov) [reply app/middleware/auth_redirect_middleware.rb]: yes, in the root folder
`lib/rack/auth_redirect` if you want to keep the naming Lib::Rack::AuthRedirect

File: app/views/index.html.haml [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+                %select.custom-select{:name => "difficulty", :required => ""}
+                  %option{:value => ""}= I18n.t(:difficulty_select_default)
+                  %option{:value => "kid"} #{DIFFICULTIES.keys[0]} - #{DIFFICULTIES[:kid][:tries]} attempts - #{DIFFICULTIES[:kid][:hints]} hints
+                  %option{:value => "average"} #{DIFFICULTIES.keys[1]} - #{DIFFICULTIES[:average][:tries]} attempts - #{DIFFICULTIES[:average][:hints]} hints
+                  %option{:value => "hacker"} #{DIFFICULTIES.keys[2]} - #{DIFFICULTIES[:kid][:tries]} attempts - #{DIFFICULTIES[:hacker][:hints]} hints
+              %hr.mb-4/
+              %button.btn.btn-primary.btn-lg{:type => "submit"}= I18n.t(:start_game_button)
+              %span.float-right
+                %a.btn.btn-warning.btn-lg{:href => "/rules", :role => "button"}= I18n.t(:rules_button)
+                %a.btn.btn-success.btn-lg{:href => "/statistics", :role => "button"}= I18n.t(:statistics_button)
```
Comment (yaorlov) [inline app/views/index.html.haml]: no new line

**Review by yaorlov — APPROVED** [review]

**Pull request merged**